### DISCUSSION TITLE
SvelteKit without SSR / SSG?

### DISCUSSION THREAD

**Discussion by mossaiby** [sveltejs/kit#3365] [Q&A] [answered]
### Describe the problem

There are several great things about SvelteKit; the power of Svelte, FS-based routing, __layouts, etc. All these goodies are affected by the extremely involved process required due to the fact that SvelteKit tries to support SSR / SSG. I do not care about SEO and search engines have learned to cope with dynamically generated layouts (or have to learn). I mostly care about all the power Svelte gives, and its design philosophy, plus the goodies that are added in SvelteKit. If SvelteKit had a way to provide all of that except SSR / SSG, I am sure we were enjoying 1.0 by now. I have struggled, for whole my weekend, to locally add Bootstrap's JS bundle to a simple SvelteKit project, and failed. Even with `adapter-static` and `await import()` (mentioned by @Rich-Harris in reply to a similar question), the build process complained about `document` not being defined.

### Describe the proposed solution

Provide a way to completely get rid of all the hassle of SSR / SSG for people that do not want / need it, and simplify their lives and at the same time, increase the number of people who fall in love with Svelte / SvelteKit. Something is missing somewhere between Svelte and SvelteKit.

### Importance

i cannot use SvelteKit without it

**Comment by dummdidumm** [comment, marked as answer]
Add a `src/routes/+layout.js` file with
```js
export const ssr = false;
```

This disables SSR for all pages and you can use libraries that only work in the browser without additional checks. See the docs for more details https://kit.svelte.dev/page-options#ssr

For the adapter you can use `adapter-static` with SPA mode: https://github.com/sveltejs/kit/tree/master/packages/adapter-static#spa-mode 

`svelte.config.js` then looks something like this:

```js
import adapter from '@sveltejs/adapter-static';

/** @type {import('@sveltejs/kit').Config} */
const config = {
  kit: {
    adapter: adapter({
      fallback: 'index.html',
    }),

    prerender: {
      entries: [],
    },
};

export default config;
```

**Reply by rdela** [reply]
as [mrkishi](https://github.com/mrkishi) wrote in https://github.com/sveltejs/kit/issues/4849#issuecomment-1121710739

> fallback: 'index.html' will clash with your actual routes/index.svelte route

more discussion/links here https://github.com/sveltejs/kit/issues/1503#issuecomment-1160838879

**Reply by OfficialCRUGG** [reply]
If I set ssr to false on the root layout, I just get a blank page...

**Reply by mossaiby** [reply]
*A lot* of things has changed since this discussion... please consult the documentation.

**Reply by OfficialCRUGG** [reply]
Actually, setting ssr to false in the root layout is exactly the solution the docs still have: https://kit.svelte.dev/docs/single-page-apps
But exactly that makes my page go blank. If I don't do it, I have a perfectly working SPA on build, but not on dev server.

**Comment by mossaiby** [comment]
Thanks @dummdidumm, I will try the solution. Two minor issues: This should be more visible, as many people may need it, and, it is so much better to have a minimal example somewhere, so people can base their work on.

**Comment by mossaiby** [comment]
Still the same issue:
In `__layout.svelte`:
```html
<script>
  import 'bootstrap';
</script>
```
In `src/hooks.js`:
```js
export const handle = async function handle({ request, resolve }) {
  const response = await resolve(request, {
    ssr: false
  });
  return response;
}
```
In `svelte.config.js`:
```js
import adapter from "@sveltejs/adapter-static";

/** @type {import('@sveltejs/kit').Config} */
const config = {
  kit: {
    adapter: adapter({
      fallback: 'index.html',
    }),

    // hydrate the <div id="svelte"> element in src/app.html
    target: '#svelte',
  },
};

export default config;
```
Now, `npm run build` gives:
```
> Using @sveltejs/adapter-static
...
ReferenceError: document is not defined
    at enableDismissTrigger (/home/admin/qplot/src/frontend/node_modules/bootstrap/dist/js/bootstrap.js:761:21)
    at /home/admin/qplot/src/frontend/node_modules/bootstrap/dist/js/bootstrap.js:856:3
    at /home/admin/qplot/src/frontend/node_modules/bootstrap/dist/js/bootstrap.js:7:83
    at Object.<anonymous> (/home/admin/qplot/src/frontend/node_modules/bootstrap/dist/js/bootstrap.js:10:3)
    at Module._compile (node:internal/modules/cjs/loader:1109:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1138:10)
    at Module.load (node:internal/modules/cjs/loader:989:32)
    at Function.Module._load (node:internal/modules/cjs/loader:829:14)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/translators:201:29)
    at ModuleJob.run (node:internal/modules/esm/module_job:175:25)
ReferenceError: document is not defined
```
Any idea?

**Reply by Badlapje** [reply]
> In `src/hooks.js`:
> 
> ```js
> export const handle = async function handle({ request, resolve }) {
>   const response = await resolve(request, {
>     ssr: false
>   });
>   return response;
> }
> ```

This bit worked for me, thanks @mossaiby

**Reply by mossaiby** [reply]
@Badlapje You're welcome!

FYI, if you upgrade your SvelteKit version, you'll need this instead:
```js
export const handle = async ({ event, resolve }) => {
  const response = await resolve(event, {
    ssr: false,
  });
  return response;
};
```

**Comment by mossaiby** [comment]
The following `svelte.config.js` solves the issue (thanks to #1650):
```js
import adapter from '@sveltejs/adapter-static';

/** @type {import('@sveltejs/kit').Config} */
const config = {
  kit: {
    adapter: adapter({
      fallback: 'index.html',
    }),

    prerender: {
      enabled: false,
    },

    ssr: false,

    // hydrate the <div id="svelte"> element in src/app.html
    target: '#svelte',
  },
};

export default config;
```
I am not sure if `ssr: false` is in fact required. Anyway, I still think that several things (documentation included) must be simplified for this use case; maybe 'Creating a SPA' should become an option when creating a new SvelteKit app?

Thanks @dummdidumm for helping out!

**Comment by mossaiby** [comment]
@dummdidumm I am getting this now when running `npm run dev`, while `npm run build` is working fine:
```
10:56:10 AM [vite] Error when evaluating SSR module /src/routes/__layout.svelte:
ReferenceError: document is not defined
    at enableDismissTrigger (/home/admin/qplot/frontend/node_modules/bootstrap/dist/js/bootstrap.js:761:21)
    at /home/admin/qplot/frontend/node_modules/bootstrap/dist/js/bootstrap.js:856:3
    at /home/admin/qplot/frontend/node_modules/bootstrap/dist/js/bootstrap.js:7:83
    at Object.<anonymous> (/home/admin/qplot/frontend/node_modules/bootstrap/dist/js/bootstrap.js:10:3)
    at Module._compile (node:internal/modules/cjs/loader:1109:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1138:10)
    at Module.load (node:internal/modules/cjs/loader:989:32)
    at Function.Module._load (node:internal/modules/cjs/loader:829:14)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/translators:201:29)
    at ModuleJob.run (node:internal/modules/esm/module_job:175:25)
document is not defined
```
It seems that the locally served site works fine, though. Any idea? This is getting really annoying...

**Reply by dummdidumm** [reply]
Are you using the latest version of SvelteKit? This feature is very new.

**Reply by mossaiby** [reply]
You are probably right... I did a `npm update` and a few packages were updated. Now I am using `SvelteKit v1.0.0-next.231` and running `npm run dev` gave me an error about `config.kit.ssr` being removed. For now everything is working fine! Thanks again, @dummdidumm!

**Comment by BillBuilt** [comment]
Hello,
I've been looking in to the SPA Mode and I'm really liking the idea, but still left wondering.... what is the fallback page actually supposed to contain? I have found no reference to its actual contents.

From the adapter-static github page: 
```
The fallback page is a blank HTML page that loads your SvelteKit app and navigates to the correct route. For example Surge, a static web host, lets you add a 200.html file that will handle any requests that don't otherwise match. 
```

This seems a bit vague to me and not sure how it would apply to SvelteKit specifically. In the Surge example it is simply showing the HTML file as being: 
```
<!-- 200.html -->
<script src="app.js"></script>
```

What would the SvelteKit equivalent be?

Thanks!

**Reply by dummdidumm** [reply]
Please open a separate discussion for this, the original question has been answered.

**Reply by mossaiby** [reply]
I was wondering for a long time about this; I tried to take it *literally* blank to start, and to my surprise, the fallback page IS CREATED for you by the `adapter-static`; it is not an input file you have to create! The documentation is quite cryptic on this subject. Do you want another surprise? Try NOT setting ` prerender: { enabled: false }`! You will receive an identical fallback page in every place you have a route (`.svelte` file). While being messy, it is surprisingly useful; you won't need special SPA settings on your web server to avoid 404 error when the user starts from anything but the `/` route.
@dummdidumm I will reproduce my answer if @BillBuilt opens another discussion.

**Reply by BillBuilt** [reply]
@mossaiby Thanks for that info! I will play around with it as soon as I can get a chance to.

@dummdidumm New discussion created at https://github.com/sveltejs/kit/issues/3413

**Reply by mossaiby** [reply]
@BillBuilt You're welcome!

@dummdidumm I also duplicated my answer there, as promised.

**Comment by cuzzea** [comment]
This information is already obsolete? I followed this discussion and nothing worked :)

IMHO not having a simple option to roll out html like vanila svelt has and decouple server from backend is the most sad thing about sveltkit as I won't be able to use it. And I bet there are others in the same situation as serverless is more and more used.

**Reply by dummdidumm** [reply]
I updated the answer to match the latest version of SvelteKit

**Comment by tonysepia** [comment]
```
 prerender: {
      enabled: false,
    },

    ssr: false,

    // hydrate the <div id="svelte"> element in src/app.html
    target: '#svelte',
```

the prerender, ssr and target parameters are causing the npm run dev to produce an error:

```
> vite dev

error when starting dev server:
Error: Unexpected option config.kit.ssr
```

Is this still up-to-date? If not, is there a way to hide this from search engine indexers?

**Reply by mossaiby** [reply]
As far as I know, things have DRASTICALLY have changed since this was written. See the docs for more information.

**Comment by RoyRao2333** [comment]
@dummdidumm @mossaiby 

After reading this thread's replies, I'm still puzzled by a few things:

1. I want to turn off SSR for my entire app, and I don't know which one is correct in the latest SvelteKit:
    - [In the tutorial](https://learn.svelte.dev/tutorial/ssr) it tells me to create `+layout.server.js` and set `export const ssr = false`.
    - [In the docs](https://kit.svelte.dev/docs/page-options#ssr) it tells me to create `+layout.ts` and do the rest.
2. What's the main difference between `@sveltejs/adapter-auto` and `@sveltejs/adapter-static`? I create my project via `pnpm create svelte` and it initially used `import adapter from '@sveltejs/adapter-auto'`. If I wanna turn off SRR completely, should I use `@sveltejs/adapter-static` instead of `@sveltejs/adapter-auto`?
3. If it doesn't bother you, could you explain what `routes/+*.js` for? Does that mean every time I wanna configure a page called `+abc.svelte`, I need to create a corresponding `+abc.js` file?

Much thanks!

**Reply by OfficialCRUGG** [reply]
I'm still quite new to Svelte & SvelteKit, but I'm also developing a SPA app right now and here's what I found out so far:
1. Set `export const ssr = false;` in `+layout.js`/`+layout.ts`, depending on what you're using.
2. Not entirely sure on the exact technical difficulties, but if you want 100% clientside rendering with no serverside, go with `adapter-static`
3. That file appears to be for metadata (like setting ssr, static generation, etc. for a page) as well as running some code to "prepare" data to put into the page from what it seems? Basically, if you come across something that needs that file (I think accessing dynamic url params needs it), create it, otherwise, no need.

**Reply by RoyRao2333** [reply]
@OfficialCRUGG 

Appreciated for the help! 😋 What really bothers me is

> difference between @sveltejs/adapter-auto and @sveltejs/adapter-static

And here's the kicker: after all the maneuvering you've done, still cannot be sure you've achieved *100% clientside rendering with no serverside* unless we're given a direct official option for `ssr: false`.

**Reply by infovore** [reply]
`adapter-auto` = detects based on environment; essentially, automatically chooses from netlify/cloudflare/vercel just by you deploying there. And what those adapters will do is serve client-side as static files, but convert any server-side routes to cloud functions for the provider of choice (assuming ssr isn't disabled for those routes).

`adapter-static` = bake client-side as static html, and any server-side data is baked at point of render. Essentially, a genuine static-site generator. So if you have a route that depends on a `.server.ts` file, that route will have both its front-end HTML _and_ the JSON that generates it baked out. (Unless, of course, you disable CSR, in which case you won't get the JSON).

**Reply by RoyRao2333** [reply]
@infovore 

Thanks for the explanation! So according to your saying

> those adapters will do is serve client-side as static files, but convert any server-side routes to cloud functions for the provider of choice

If I'm using `adapter-auto` and not using any of those cloud services, when I set `export const ssr = false` in `+layout.ts`, does this mean my app is a pure client-side app with no SSR feature and need no further configurations?

**Reply by infovore** [reply]
I believe so; to quote the documentation:

> If you add `export const ssr = false` to your root `+layout.ts`, your entire app will only be rendered on the client — which essentially means you turn your app into an SPA.

Bear in mind that _in development_ your application is essentially running using `adapter-node` - a Node/Express server that serves everything up, client-and-server side. Again, this can be forced into client-side only mode as above. The adapters only kick in when you `build` the site for deployment.

In general: if you are planning to just build an entirely client-side pile of HTML, I'd suggest using the most specific adapter for the job, which is `adapter-static`.

**Reply by RoyRao2333** [reply]
@infovore 

Great explanations! 

In fact, I need my app to be client-side only because I'm concerned about using `document`, `window` or anything else that only exist in a browser. I've met quite bugs with Next.js SSR because lots of third-party components and utils are not compatible with SSR.

**Reply by infovore** [reply]
I'll note that this isn't next.js, so you shouldn't make decisions about SvelteKit based on it.

There's a useful variable called `browser` to be imported from `$app/environment' - [see the docs](https://kit.svelte.dev/docs/modules#$app-environment-browser).

Essentially, that means that in a component, you can do the following:

```js
import { browser } from '$app/environment;
import { onMount } from 'svelte'

onMount(() => {
  // things that should happen when the component mounts in any environment
  
  if(browser) {
    document.doSomethingOnlyAvailableInADOM();
  }
})
```

This allows a component to have browser-specific functionality without requiring the whole app just assumes it's all one big SPA.

**Reply by RoyRao2333** [reply]
@infovore 

Wow, I didn't notice this feature before and I'm just learning Svelte for a few days.

Still, thanks for pointing it out for me!
